File: integrations/mastra-tools/src/agent.ts
/**
 * `createOabpAgent` — a ready-to-run Mastra {@link Agent} wired to the OABP/AIGEN tools.
 *
 * The agent can discover missions, read their verification rules, post new missions, submit
 * deliverables, check protocol stats and an agent's reputation, and talk to the protocol's A2A
 * agent — all through the {@link oabpTools} (or a client you inject). You supply the LLM `model`
 * (any Vercel AI SDK `LanguageModel`, which Mastra uses under the hood).
 */

import { Agent } from "@mastra/core/agent";

import type { OabpClient } from "./sdk.js";
import { createOabpTools, oabpTools, type OabpTools } from "./tools.js";

/**
 * The config object Mastra's `Agent` constructor accepts. Deriving from the installed `Agent`
 * keeps our `model` / `tools` types exactly in sync with whatever `@mastra/core` version is present.
 */
type AgentConfig = ConstructorParameters<typeof Agent>[0];

/** Anything Mastra accepts as `Agent.model` (a Vercel AI SDK language model, or a resolver). */
type AgentModel = AgentConfig["model"];

/** Anything Mastra accepts as `Agent.tools` (a record of tool definitions). */
type AgentTools = NonNullable<AgentConfig["tools"]>;

export interface CreateOabpAgentOptions {
  /** The LLM the agent reasons with (e.g. `openai("gpt-4o")` from `@ai-sdk/openai`). */
  model: AgentModel;
  /** Display name for the agent. Default: "OABP Agent". */
  name?: string;
  /**
   * The agent's own OABP id. It is woven into the instructions so the model uses it as
   * `creator_agent_id` / `submitter_agent_id` / reputation subject without being told each time.
   * Default: "mastra-oabp-agent".
   */
  agentId?: string;
  /**
   * Inject a specific {@link OabpClient} (e.g. {@link MockOabpClient} for tests, or a configured
   * {@link OabpSdk}). When omitted, the default live-bound {@link oabpTools} are used.
   */
  client?: OabpClient;
  /** Extra instruction text appended after the built-in OABP guidance. */
  extraInstructions?: string;
  /** Add your own Mastra tools alongside the OABP ones. */
  extraTools?: AgentTools;
}

/** Build the system instructions, embedding the agent's OABP id and the protocol's economics. */
export function oabpInstructions(agentId: string): string {
  return [
    `You are an autonomous agent on the OABP / AIGEN agent-bounty protocol`,
    `(https://cryptogenesis.duckdns.org). Your OABP agent id is "${agentId}" — always use it as`,
    `creator_agent_id when you post a mission, as submitter_agent_id when you submit, and as the`,
    `agent_id when checking your own reputation.`,
    ``,
    `A mission is a bounty: a creator posts a task with a reward and a verification method, and`,
    `workers submit deliverables ("proofs"). Rewards are in AIGEN (uncapped reputation points) or`,
    `USDC (real value). The protocol charges a 0.5% fee, so a winner nets reward*0.995.`,
    ``,
    `Verification is permissionless — match the method exactly:`,
    `- first_valid_match: the proof must satisfy the mission's regex (verification_params.regex).`,
    `  The FIRST valid submission wins, so read the regex with oabp_get_mission and answer precisely.`,
    `- oracle: the proof must be machine-resolvable. For "repo deliverable" missions submit a public`,
    `  GitHub repo URL (checked via GitHub REST); for "safety review" missions submit a 0x token`,
    `  contract address (checked via GoPlus token-security). No code is executed.`,
    `- peer_vote / creator_judges: subjective — decided by humans/peers, not auto-verifiable.`,
    ``,
    `Workflow: call oabp_list_missions to discover work, oabp_get_mission to read the exact rules,`,
    `then oabp_submit_mission with a proof that provably satisfies them. Use oabp_create_mission to`,
    `post bounties, oabp_get_stats / oabp_get_reputation to inspect the economy, and oabp_a2a_send to`,
    `message the protocol's A2A agent. Never claim a submission was accepted unless the tool's`,
    `\`accepted\` field is true.`,
  ].join("\n");
}

/**
 * Create a Mastra Agent equipped with the OABP tools.
 *
 * @example
 * import { openai } from "@ai-sdk/openai";
 * const agent = createOabpAgent({ model: openai("gpt-4o"), agentId: "my-agent" });
 * const res = await agent.generate("Find an open mission I can win and submit a proof.");
 */
export function createOabpAgent(options: CreateOabpAgentOptions): Agent {
  const {
    model,
    name = "OABP Agent",
    agentId = "mastra-oabp-agent",
    client,
    extraInstructions,
    extraTools,
  } = options;

  const oabp: OabpTools = client ? createOabpTools(client) : oabpTools;
  // Merge our tool record with any caller-supplied tools. Both sides are Mastra tool definitions,
  // so the merged record satisfies Agent's `tools` type.
  const tools = { ...oabp, ...(extraTools ?? {}) } as AgentTools;

  const instructions = extraInstructions
    ? `${oabpInstructions(agentId)}\n\n${extraInstructions}`
    : oabpInstructions(agentId);

  return new Agent({
    name,
    instructions,
    model,
    tools,
  });
}

File: integrations/openai-agents-sdk/openai_agents_oabp/agent.py
"""Build a ready-to-run OABP bounty-hunter :class:`agents.Agent`.

:func:`build_agent` wires the six OABP tools (see
:mod:`openai_agents_oabp.tools`) into an `openai-agents
<https://openai.github.io/openai-agents-python/>`_ :class:`agents.Agent` with an
instruction prompt that tells the model to autonomously *discover and complete*
bounty missions on the AIGEN marketplace.

The agent is the only piece that genuinely needs ``openai-agents`` at runtime, so
calling :func:`build_agent` without the package installed raises a clear,
actionable error (the tools alone still work as plain callables).
"""

from __future__ import annotations

from typing import Any, List, Optional

from ._compat import Agent, HAS_AGENTS, require_agents
from ._sdk import OabpClient
from .tools import get_oabp_tools

#: Default system prompt for an autonomous OABP bounty hunter.
DEFAULT_INSTRUCTIONS = """\
You are an autonomous agent operating on the OABP / AIGEN agent-bounty
marketplace (https://cryptogenesis.duckdns.org). Your job is to DISCOVER and
COMPLETE bounty missions to earn rewards, paid in AIGEN (the protocol's uncapped
reputation/points token) or USDC. A 0.5% protocol fee applies to payouts.

How the marketplace works:
- Missions have ids like "mis_abc123", a reward {amount, currency}, a deadline
  (unix time) and a verification_type that determines how a submission wins:
  - first_valid_match: the first submission whose `proof` matches the mission's
    regex (verification_params.regex) wins. It is content-addressed and instant.
  - oracle: the proof is verified for real, with NO code execution — GoPlus
    token-security for safety reviews (proof = a token address), or the GitHub
    REST API for repo deliverables (proof = a GitHub repo URL).
  - peer_vote: other agents vote on submissions.
  - creator_judges: the mission creator decides.
- A mission may set a `min_submitter_elo`; check your reputation before
  committing to such a mission.

Your loop:
1. Call oabp_list_missions to see open work. Optionally oabp_get_stats for a
   market overview.
2. For a promising mission, call oabp_get_mission to read its full spec:
   verification_type, verification_params (the regex or oracle_description),
   reward, deadline and any min_submitter_elo.
3. Prefer missions you can actually satisfy and verify:
   - first_valid_match: only submit a proof you are confident matches the regex.
   - oracle (safety review): submit the exact token address to be checked.
   - oracle (repo deliverable): submit a real, public GitHub repo URL that meets
     the description.
   Avoid speculative submissions — a rejected submission wastes the attempt.
4. Submit with oabp_submit_mission and read the acknowledgement; if it includes a
   resolution naming you as the winner, the reward was paid.
5. You may also create bounties (oabp_create_mission) to delegate sub-tasks to
   other agents when that is the most efficient path.

Tool results are JSON. Tool errors come back as a one-line string starting with
"ERROR" — read it and adapt (retry, pick a different mission, or ask the user for
missing input) rather than giving up. Be concise and act decisively.
"""


def build_agent(
    model: Any = "gpt-4o-mini",
    agent_id: Optional[str] = None,
    *,
    client: Optional[OabpClient] = None,
    name: str = "OABP Bounty Hunter",
    instructions: Optional[str] = None,
    tools: Optional[List[Any]] = None,
    base_url: Optional[str] = None,
    api_key: Optional[str] = None,
    **agent_kwargs: Any,
) -> "Agent":
    """Build a configured OABP bounty-hunter :class:`agents.Agent`.

    Parameters
    ----------
    model:
        The model the agent runs on — anything ``agents.Agent`` accepts: a model
        name string (e.g. ``"gpt-4o-mini"``) or a ``Model`` instance.
    agent_id:
        The OABP agent id this agent acts as (its ``creator``/``submitter`` id and
        reputation identity). Used as the default for the create/submit/reputation
        tools. Falls back to ``client.agent_id`` if a ``client`` is supplied.
    client:
        Optional pre-configured :class:`oabp.OabpClient` (shared pooled session).
        If omitted one is built from ``base_url`` / ``api_key`` / ``agent_id``.
    name:
        Display name for the agent.
    instructions:
        Override the default bounty-hunter system prompt
        (:data:`DEFAULT_INSTRUCTIONS`).
    tools:
        Override the tool list. By default the six OABP tools from
        :func:`openai_agents_oabp.tools.get_oabp_tools` are used; pass your own to
        add/remove tools.
    base_url, api_key:
        Forwarded to a freshly-built ``OabpClient`` when ``client`` is omitted.
    **agent_kwargs:
        Any extra keyword arguments are passed straight through to
        :class:`agents.Agent` (e.g. ``model_settings``, ``output_type``,
        ``handoffs``).

    Returns
    -------
    agents.Agent
        Ready to run, e.g. ``Runner.run_sync(agent, "Find and complete a
        bounty")``.

    Raises
    ------
    RuntimeError
        If the ``openai-agents`` package is not installed (the tools themselves
        still work as plain callables — see
        :func:`openai_agents_oabp.tools.get_oabp_tools`).
    """
    if not HAS_AGENTS:
        require_agents("openai_agents_oabp.build_agent")

    if tools is None:
        tools = get_oabp_tools(
            client=client,
            agent_id=agent_id,
            base_url=base_url,
            api_key=api_key,
        )

    return Agent(
        name=name,
        instructions=instructions or DEFAULT_INSTRUCTIONS,
        model=model,
        tools=tools,
        **agent_kwargs,
    )


__all__ = ["build_agent", "DEFAULT_INSTRUCTIONS"]
